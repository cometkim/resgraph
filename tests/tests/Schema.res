/** A user in the system. */
@gql.type
type user = {
  /** The users name.*/ name: string,
  @gql.field /** The age of the user. */
  age: int,
  @gql.field @deprecated("Use 'age' instead.") /** The last age of the user. */
  lastAge: option<int>,
}

/** A group in the system. */
@gql.type
type group = {
  /** The group name.*/
  @gql.field
  name: string,
  memberIds: array<string>,
}

/** A user or a group. */
@gql.union
type userOrGroup = | /** This is a user.*/ User(user) | /** And this is a group. */ Group(group)

/** Indicates what status a user currently has. */
@gql.enum
type userStatus =
  | /** User is online. */ Online
  | /** User is offline. */ Offline
  | /** User is idle. */
  @deprecated("Use 'Offline' instead.")
  Idle

module UserFields = {
  @gql.field
  let id = (user: user) => {
    ("User:" ++ user.name)->ResGraph.id
  }

  @gql.field
  let name = (user: user, ~includeFullName) => {
    let includeFullName = includeFullName->Belt.Option.getWithDefault(false)

    if includeFullName {
      user.name
    } else {
      user.name->Js.String2.slice(~from=0, ~to_=3)
    }
  }

  @gql.field
  let currentStatus = (_user: user) => {
    Online
  }

  @gql.field
  let allNames = (user: user) => {
    [user.name]
  }
}

@gql.type
type query = {}

@gql.inputObject /** Configuration for searching for a user.*/
type userConfig = {
  /** The ID of a user to search for. */ id: string,
  /** The name of the user to search for. */
  @deprecated("This is going away")
  name?: string,
}

module QueryFields = {
  @gql.field
  let me = (_: query): option<user> => {
    Some({name: "Hello", age: 35, lastAge: None})
  }

  @gql.field
  let entity = (_: query, ~id: ResGraph.id, ~ctx: ResGraphContext.context) => {
    ignore(id)
    ignore(ctx)
    User({name: "Hello", age: 35, lastAge: None})
  }

  @gql.field
  let searchForUser = (_: query, ~input: userConfig): option<user> => {
    Some({name: input.name->Belt.Option.getWithDefault("Hello"), age: 35, lastAge: None})
  }

  @gql.field
  let allowExplicitNull = (_: query, ~someNullable) => {
    let wasNull = someNullable == Js.Nullable.null
    let wasUndefined = someNullable == Js.Nullable.undefined

    if wasNull {
      "Was null"
    } else if wasUndefined {
      "Was undefined"
    } else {
      someNullable->Js.Nullable.toOption->Belt.Option.getExn
    }
  }

  @gql.field
  let listAsArgs = (
    _: query,
    ~regularList,
    ~optionalList=?,
    ~nullableList,
    ~nullableInnerList,
    ~list1: option<array<option<string>>>,
    ~list2: option<array<option<array<option<string>>>>>,
    ~list3: option<array<option<array<array<string>>>>>,
  ) => {
    ignore(list1)
    ignore(list2)
    ignore(list3)
    let regularList = regularList->Belt.Array.keepMap(v => v)
    let optionalList = optionalList->Belt.Option.getWithDefault([])
    let nullableList =
      nullableList->Js.Nullable.toOption->Belt.Option.getWithDefault([])->Belt.Array.keepMap(v => v)
    let nullableInnerList =
      nullableInnerList
      ->Js.Nullable.toOption
      ->Belt.Option.getWithDefault([])
      ->Belt.Array.keepMap(Js.Nullable.toOption)

    let arr =
      Belt.Array.concatMany([
        regularList,
        optionalList,
        nullableList,
        nullableInnerList,
      ])->Js.Array2.map(str => "v " ++ str)

    arr
  }
}

// ^gen
