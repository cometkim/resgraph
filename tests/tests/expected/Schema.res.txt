generating schema from tests/Schema.res

Adding enum userStatus
Found potential resolver fn: id
Checking for type to attach resolver to
user
already seen User
ResGraph.id
Found potential resolver fn: name
Checking for type to attach resolver to
user
already seen User
Found potential resolver fn: currentStatus
Checking for type to attach resolver to
user
already seen User
userStatus
matched enum and variant
Adding enum UserStatus
Found potential resolver fn: me
Checking for type to attach resolver to
query
already seen Query
user
already seen User
@@warning("-27")

open ResGraph__GraphQLJs

let enum_UserStatus = GraphQLEnumType.make({
  name: "UserStatus",
  description: " Indicates what status a user currently has. ",
  values: {
    "Online": {
      GraphQLEnumType.value: "Online",
      description: " User is online. ",
      deprecationReason: ?None,
    },
    "Offline": {
      GraphQLEnumType.value: "Offline",
      description: " User is offline. ",
      deprecationReason: ?None,
    },
    "Idle": {
      GraphQLEnumType.value: "Idle",
      description: " User is idle. ",
      deprecationReason: "Use 'Offline' instead.",
    },
  }->makeEnumValues,
})
let enum_userStatus = GraphQLEnumType.make({
  name: "userStatus",
  description: " Indicates what status a user currently has. ",
  values: {
    "Online": {
      GraphQLEnumType.value: "Online",
      description: " User is online. ",
      deprecationReason: ?None,
    },
    "Offline": {
      GraphQLEnumType.value: "Offline",
      description: " User is offline. ",
      deprecationReason: ?None,
    },
    "Idle": {
      GraphQLEnumType.value: "Idle",
      description: " User is idle. ",
      deprecationReason: "Use 'Offline' instead.",
    },
  }->makeEnumValues,
})
let t_User: ref<GraphQLObjectType.t> = Obj.magic({"contents": Js.null})
let get_User = () => t_User.contents
let t_Query: ref<GraphQLObjectType.t> = Obj.magic({"contents": Js.null})
let get_Query = () => t_Query.contents

t_User.contents = GraphQLObjectType.make({
  name: "User",
  fields: () =>
    {
      "currentStatus": {
        typ: enum_UserStatus->GraphQLEnumType.toGraphQLType->nonNull,
        resolve: makeResolveFn((src, args, ctx) => {Schema.UserFields.currentStatus(src)}),
      },
      "name": {
        typ: Scalars.string->Scalars.toGraphQLType->nonNull,
        args: {"includeFullName": {typ: Scalars.boolean->Scalars.toGraphQLType}}->makeArgs,
        resolve: makeResolveFn((src, args, ctx) => {
          Schema.UserFields.name(
            src,
            ~includeFullName=args["includeFullName"]->Js.Nullable.toOption,
          )
        }),
      },
      "id": {
        typ: Scalars.id->Scalars.toGraphQLType->nonNull,
        resolve: makeResolveFn((src, args, ctx) => {Schema.UserFields.id(src)}),
      },
      "age": {
        typ: Scalars.int->Scalars.toGraphQLType->nonNull,
        resolve: makeResolveFn((src, _args, _ctx) => src["age"]),
      },
    }->makeFields,
})
t_Query.contents = GraphQLObjectType.make({
  name: "Query",
  fields: () =>
    {
      "me": {
        typ: get_User()->GraphQLObjectType.toGraphQLType,
        resolve: makeResolveFn((src, args, ctx) => {Schema.QueryFields.me(src)}),
      },
    }->makeFields,
})

let schema = GraphQLSchemaType.make({"query": get_Query()})


